fem/dfem/doperator.hpp

```diff
@@ -0,0 +1,698 @@
+/// automatic differentiation to compute derivatives with respect to solutions
+/// (Jacobians) and parameter fields (general derivative operators).
+///
+/// The operator is constructed with solution fields that it will act on and
+/// parameter fields that define coefficients. Quadrature functions are added by
+/// e.g. using AddDomainIntegrator() which specify how the operator evaluates f
+/// those functionas and parameters at quadrature points.
+///
+/// Derivatives can be computed by obtaining a DerivativeOperator using
+/// GetDerivative().
+///
+/// @see DerivativeOperator
+class DifferentiableOperator : public Operator
+{
+public:
+   /// Constructor for the DifferentiableOperator class.
+   ///
+   /// @param solutions The solution fields that the operator will act on.
+   /// @param parameters The parameter fields that define coefficients.
+   /// @param mesh The mesh on which the operator is defined.
+   DifferentiableOperator(
+      const std::vector<FieldDescriptor> &solutions,
+      const std::vector<FieldDescriptor> &parameters,
+      const ParMesh &mesh);
+
+   /// @brief Compute the action of the operator on a given vector.
+   ///
+   /// @param solutions_t The solution vector in which to compute the action. This has to be a T-dof vector.
+   /// @param result_t Result vector of the action of the operator on
+   /// solutions_t. The result is a T-dof vector.
+   void Mult(const Vector &solutions_t, Vector &result_t) const override
+   {
+      MFEM_ASSERT(!action_callbacks.empty(), "no integrators have been set");
+      prolongation(solutions, solutions_t, solutions_l);
+      for (auto &action : action_callbacks)
+      {
+         action(solutions_l, parameters_l, residual_l);
+      }
+      prolongation_transpose(residual_l, result_t);
+   }
+
+   /// @brief Add a domain integrator to the operator.
+   ///
+   /// @param qfunc The quadrature function to be added.
+   /// @param inputs Tuple of FieldOperators for the inputs of the quadrature
+   /// function.
+   /// @param outputs Tuple of FieldOperators for the outputs of the quadrature
+   /// function.
+   template <
+      typename func_t,
+      typename... input_ts,
+      typename... output_ts,
+      typename derivative_ids_t = decltype(std::make_index_sequence<0> {})>
+   void AddDomainIntegrator(
+      func_t &qfunc,
+      tuple<input_ts...> inputs,
+      tuple<output_ts...> outputs,
+      const IntegrationRule &integration_rule,
+      const Array<int> domain_attributes,
+      const derivative_ids_t derivative_ids = std::make_index_sequence<0> {});
+
+   /// @brief Set the parameters for the operator.
+   ///
+   /// This has to be called before using Mult() or MultTranspose().
+   ///
+   /// @param p The parameters to be set. This should be a vector of pointers to
+   /// the parameter vectors. The vectors have to be L-vectors (e.g.
+   /// GridFunctions).
+   void SetParameters(std::vector<Vector *> p) const;
```

Comment:
We can consider turning this to `const std::vector<Vector *> &`

Reply:
No objections. It's not going to buy us much though.